pkg/db/compress/compress.go

```diff
@@ -0,0 +1,246 @@
+	opts.zstdCompressionLevel = int(o)
+}
+
+func WithZstdCompressionLevel(level int) Option {
+	return zstdCompressionLevelOption(level)
+}
+
+type zstdCompressionThreadsOption int
+
+func (o zstdCompressionThreadsOption) apply(opts *options) {
+	opts.zstdCompressionThreads = int(o)
+}
+
+func WithZstdCompressionThreads(threads int) Option {
+	return zstdCompressionThreadsOption(threads)
+}
+
+type debugOption bool
+
+func (o debugOption) apply(opts *options) {
+	opts.debug = bool(o)
+}
+
+func WithDebug(debug bool) Option {
+	return debugOption(debug)
+}
+
+func Compress(opts ...Option) error {
+	options := &options{
+		dbtype:                 "boltdb",
+		dbpath:                 filepath.Join(utilos.UserCacheDir(), "vuls.db"),
+		boltdbNoSync:           false,
+		boltdbTxMaxSize:        65536,
+		zstdCompressionLevel:   22,
+		zstdCompressionThreads: runtime.NumCPU(),
+		debug:                  false,
+	}
+	for _, o := range opts {
+		o.apply(options)
+	}
+
+	switch options.dbtype {
+	case "boltdb":
+		if err := compactBoltDB(options.dbpath, options.boltdbNoSync, options.boltdbTxMaxSize); err != nil {
+			return errors.Wrap(err, "compact boltdb")
+		}
+		if err := compressZstandard(options.dbpath, options.zstdCompressionLevel, options.zstdCompressionThreads); err != nil {
+			_ = os.Remove(fmt.Sprintf("%s.zst", options.dbpath))
+			return errors.Wrap(err, "compress db with zstd")
+		}
+	default:
+		return errors.Errorf("%s is not support dbtype", options.dbtype)
+	}
+
+	return nil
+}
+
+func compactBoltDB(dbpath string, noSync bool, txMaxSize int64) error {
+	si, err := os.Stat(dbpath)
+	if err != nil {
+		return errors.Wrapf(err, "stat %q", dbpath)
+	}
+
+	src, err := bbolt.Open(dbpath, 0400, &bbolt.Options{ReadOnly: true})
+	if err != nil {
+		return errors.Wrapf(err, "open %q", dbpath)
+	}
+	defer src.Close() //nolint:errcheck
+
+	tmpdb, err := func() (string, error) {
+		f, err := os.CreateTemp(filepath.Dir(dbpath), fmt.Sprintf("%s.*", filepath.Base(dbpath)))
+		if err != nil {
+			return "", errors.Wrapf(err, "create %q", filepath.Join(filepath.Dir(dbpath), fmt.Sprintf("%s.*", filepath.Base(dbpath))))
+		}
+		defer f.Close() //nolint:errcheck
+		return f.Name(), nil
+	}()
+	if err != nil {
+		return errors.Wrapf(err, "create temp db")
+	}
+	defer os.Remove(tmpdb) //nolint:errcheck
+
+	dst, err := bbolt.Open(tmpdb, si.Mode(), &bbolt.Options{NoSync: noSync})
+	if err != nil {
+		return errors.Wrapf(err, "open %q", tmpdb)
+	}
+	defer dst.Close() //nolint:errcheck
+
+	if err := bbolt.Compact(dst, src, txMaxSize); err != nil {
+		return errors.Wrap(err, "compact")
+	}
+
+	di, err := os.Stat(tmpdb)
+	if err != nil {
+		return errors.Wrapf(err, "stat %q", tmpdb)
+	}
+	if di.Size() == 0 {
+		return errors.Errorf("size of %q is 0", tmpdb)
+	}
+
+	sizeunit := unit(si.Size())
+	slog.Info("BoltDB compaction", "before", format(si.Size(), sizeunit), "after", format(di.Size(), sizeunit), "gain", fmt.Sprintf("%.2fx", float64(si.Size())/float64(di.Size())))
```

Comment:
This information is really nice! 🍻 
It would be great if we could see the execution times as well, isn't it?